Parameterized tests (#947)
software-mansion/protostar: website/docs/tutorials/07-testing/02-cheatcodes/example.md

# `example`

```python
def example(self, *args: Any, **kwargs: Any) -> None:
```

Parametrizes test with explicitly provided data.
You can provide multiple examples for one test.
In such a case, they are run sequentially.

> `example` can be used next to the [`given`](./given.md) cheatcode.

Reviewer: Is protostar still going to fuzz when there is no `given` cheatcode?
If yes -> this explanation doesn't capture this behaviour
If not -> behaviour is inconsistent with standard fuzz tests

Reviewer 2: > If not -> behaviour is inconsistent with standard fuzz tests

hmm you're quite right, but I wonder how to enforce doing parameterized testing without fuzzing in other way 🤔

Reviewer 2: We want to support all 4 combinations:
1. standard tests
2. fuzz-only tests
3. parameterized-only tests
4. parameterized-and-then-fuzzed tests

Each of your ideas blocks some of these 4 combos. Current API allowed everything:

1. is trivial
2. use only `given`
3. use only `example`
4. use both `example` and `given` in single test

I think maybe this isn't actually so inconsistent. If you use a cheatcode, then you explicitly ask Protostar to do run specified mode, and you can combine modes naturally.

Author: > Is protostar still going to fuzz when there is no `given` cheatcode?

No, in such a case only `examples` will be taken under consideration.

Ok, so what's your suggestion here? We discussed somewhere in the comments that if we use `example`s without `given`, then we only want to have these examples executed, unlike when we do not use either of these cheatcodes - in this case, there's going to be a default strategy for `given` provided implicitly by protostar.

I'm going to add an explicit statement about that.

Author: [add an explicit statement about using example without given](https://github.com/software-mansion/protostar/pull/947/commits/b1acac1d73d7e7a36dbfa541ac3e53e124749a77)

Author: Sorry, I couldn't see the newest responses here.

It all makes sense to me anyway. Are there any more doubts here?

Reviewer: I still have doubts being honest. We made a mistake designing initial verision of fuzzing, making it default strategy. No framework does that as far as I am aware. I would prefer to make fuzzing always explicit.

I created a issue to discuss that:

Approving, because the issue has nothing to do with your PR I believe.
Thanks for explicit statement, it makes things much cleaner.

Reviewer 2: Resolving this conversation in favour of Maks's task

In such a case, first all the `example`s are run and only then the the data from `given` is applied.

:::warning
This cheatcode is only available in [setup cases](../README.md#setup-case).
:::
:::warning
`example` is not limited by [`max_examples`](./max-examples.md) and is not connected to it in any way.
:::

```cairo title="Example"
@external
func setup_less_equal_compare() {
    %{
        example(1, 2)
        # it affects the fuzzing strategy applied after all of the examples
        given(a = strategy.integers(15, 20), b = strategy.integers(10, 14))
        example(b=6, a=5)
    %}
    return ();
}

@external

> func test_examples{syscall_ptr: felt*, range_check_ptr}(a: felt, b: felt) {# Tested against:

Reviewer: ```suggestion
func test_examples{syscall_ptr: felt*, range_check_ptr}(a: felt, b: felt) {
    # Tested against:
```

# `example`

```python

> def example(self, *args: Any, **kwargs: Any) -> None:

Reviewer: I wonder if `parametrize` wouldn't be a better name as we are already inspired by pytest heavily

Author: @mkaput What do you think? You're the original author of the issue for this feature.
`parametrize` would come from pytest
`example` from hypothesis

Reviewer 2: Pytest's `parameterize` has different API than ours. Their approach is more like `examples` (notice plural form), when you declare multiple input sets at once. Here we define inputs one by one hence I think `example` is better suit. 

also `parameters` doesn't sound as nice as `example` imho here.

Reviewer: Ok, let's leave example then

```

> This is an addition to the [`given`](./given.md) cheatcode, but, due to how the protostar works,
> `given` cheatcode does not have to be used in order to invoke `example`.
> While `given` runs tests with multiple, randomly generated data sets,
> `example` takes certain values and runs tests using them.
> Tests with the data specified in `example` will always be run before applying the data from `given`.
> 
> Multiple `example`s can be used for one test. In such a case, all of them will be run sequentially,
> only after that allowing `given` to run.
> 
> `example` is not limited by [`max_examples`](./max-examples.md) and is not connected to it in any way.
> 
> :::warning
> This cheatcode is only available in [setup cases](../README.md#setup-case).
> :::

Reviewer: I wouldn't present it as an addition to `given` cheatcode. It is a standalone feature. Please make the first sentence explanation what the cheatcode does as it is the most important information. 
I provided an simplified description, lmk what do you think

```suggestion
Parametrizes test with explicitly provided data. You can provide multiple examples for one test. In such a case, they run sequentially.

Even when examples are provided protostar still runs fuzz tests against the test case. Protostar does it after running all of the examples.

:::warning
This cheatcode is only available in [setup cases](../README.md#setup-case).
:::
:::warning
`example` is not limited by [`max_examples`](./max-examples.md) and is not connected to it in any way.
:::

```

Author: > Even when examples are provided protostar still runs fuzz tests against the test case. Protostar does it after running all of the examples.

Actually, I handled this in the latest commit and it doesn't work like that anymore. Since we want to make `example` a standalone cheatcode, it should be possible to run it without `given` and just get the examples we want. Thanks for the rest, I will come up with a new version.

Author: [update docs](https://github.com/software-mansion/protostar/pull/947/commits/75b02de8f52d7eced8f64ae8687db4947c59bb45)

```cairo title="Example"
@external
func setup_less_equal_compare() {
    %{
        example(1, 2)

>         example(2, 3)

Reviewer: ```suggestion
```

Author: [update docs](https://github.com/software-mansion/protostar/pull/947/commits/75b02de8f52d7eced8f64ae8687db4947c59bb45)

>         given(a = strategy.integers(15, 20), b = strategy.integers(10, 14))

Reviewer: ```suggestion
        # it affects the fuzzing strategy applied after all of the examples
        given(a = strategy.integers(15, 20), b = strategy.integers(10, 14))
```

        example(a=3, b=4)

>         example(b=6, a=5)

Reviewer: ```suggestion
```

Author: I'd remove the other example `example(a=3, b=4)`, I think it's good to add an example with reversed args (`b` is first, then `a`)

Author: [update docs](https://github.com/software-mansion/protostar/pull/947/commits/75b02de8f52d7eced8f64ae8687db4947c59bb45)

    %}
    return ();
}

@external
func test_examples{syscall_ptr: felt*, range_check_ptr}(a: felt, b: felt) {

>     assert_le(0, a);

Reviewer: [Nitpick] Maybe it is worth to include it:
```suggestion
    # Tested against:
    # (1,2)
    # (3,4)
    # fuzzing: a = strategy.integers(15, 20), b = strategy.integers(10, 14) 
    assert_le(0, a);
```

Author: I didn't see it in other docs (but maybe I didn't look close enough), but I don't mind adding it.

Author: [update docs](https://github.com/software-mansion/protostar/pull/947/commits/75b02de8f52d7eced8f64ae8687db4947c59bb45)
